### PR TITLE
Interrupt API

### PR DESCRIPTION
**Pull request by aykevl**
Very much incomplete, but shows how it could work.

It works locally (after patching go-llvm) and has no code size impact for blinky2 on the pca10040.

### CODE REVIEW

**Comment by aykevl** [comment]
I've taken care to produce useful error messages when an interrupt is defined twice. For example:

```
src/machine/machine_nrf.go:91:23: UART0_IRQHandler redeclared in this program
	previous declaration at src/machine/machine_nrf51.go:22
```

File: src/machine/board_trinket.go [hunk 1]
```diff
@@ -51,9 +54,8 @@ var (
 	}
 )
 
-//go:export SERCOM1_IRQHandler
-func handleUART1() {
-	defaultUART1Handler()
+func init() {
+	UART1.interrupt = interrupt.New(sam.IRQ_SERCOM0, UART1.handleInterrupt)
```
Comment (aykevl) [inline src/machine/board_trinket.go]: I found a bug here: UART1 is defined with SERCOM0 but the interrupt fired on SERCOM1. This PR fixes that.

File: src/machine/board_circuitplay_express_baremetal.go [hunk 2]
```diff
@@ -13,9 +16,8 @@ var (
 	}
 )
 
-//go:export SERCOM1_IRQHandler
-func handleUART1() {
-	defaultUART1Handler()
+func init() {
+	UART1.interrupt = interrupt.New(sam.IRQ_SERCOM4, UART1.handleInterrupt)
```
Comment (aykevl) [inline src/machine/board_circuitplay_express_baremetal.go]: Same issue here: the UART uses SERCOM4 but the interrupt fired on SERCOM1.

**Comment by deadprogram** [comment]
I just tried this branch, this was the result:

```
$ tinygo flash -target arduino-nano33 examples/serial
cannot find interrupt name for number 7
cannot find interrupt name for number 3
```

**Comment by aykevl** [comment]
You need to update the generated Go files using `make gen-device` (or specifically `make gen-device-sam`).

**Comment by deadprogram** [comment]
I should have realized that just from looking at the files changed in this PR. That did it.

I tried a few more examples that use interrupts, like UART and USB CDC and worked perfectly.

**Comment by aykevl** [comment]
This PR is now ready from my point of view.

I've done some measurements regarding code/data size changes with this PR:

  * As expected, some programs got slightly larger while others got slightly smaller. This is probably just normal optimization noise that happens due to reordered functions.
  * Programs running on the samd21 chip became slightly smaller (about 80 bytes) and consumed less RAM (144 bytes). I think this is because the `UART1` struct (including buffer) can now be optimized away because most programs don't use it.

**Comment by deadprogram** [comment]
Seems like `src/machine/machine_atsamd51.go` has the newer API for the USB and the older one for the UART? Is that intentional?

**Comment by deadprogram** [comment]
Never mind I just realized that #831 and this PR can be rebased together, and then I suppose a future PR will complete the transition.

**Comment by aykevl** [comment]
> Never mind I just realized that #831 and this PR can be rebased together, and then I suppose a future PR will complete the transition.

Exactly, that was my plan. Doing it in this PR (without #831) would complicate things.

**Comment by deadprogram** [comment]
OK I have tested with SAMD21, SAMD51, NRF52840, and AVR. So far all have worked perfectly.

Now merging, this is giant leap forward for what we can do!

**Pull request merged**